README
########################################
#
# Timebank
#
# Started in July, 2012.
# Last updated on 08/12/2012.
# Current version:	0.2
# Made and tested using python v2.6.6 in Crunchbang Statler.
#
# This is a simple, Mint-inspired time manager,
# designed to help track time spent on activities and
# display them in an easy, useful manner.
#
# This project is nowhere near finished and barely tested.
# Use "help" in python for details on individual functions.
#
# Overview:
# 
# 	Time is logged using "banks".  Banks are stored as a
# 	single file named "mybank".  It is highly recommended
# 	that you rename the bank once it's scope is over (for
# 	example, if "mybank" represents Mon - Sun, rename on
# 	next Monday and setup a new bank for that week).
#
# 	The scope of a bank isn't set in stone.  It can be
# 	changed on the fly using the "change()" command,
# 	along with the display unit and other settings
# 	(see help(change()) for usage).
#
# Usage:
#
# 	This program requires Python to be installed, preferably
# 	version 2.6.6.  Version 3 of python will likely work,
# 	but has not been tested yet (so may be buggy). This
# 	program has also been untested in anything other than *Nix
# 	(Crunchbang Statler, to be exact).
#
# 	To load Timebank, open python in the Timebank program
# 	directory, and execute "import tb" or "from tb import *".
# 	The latter is preferred, as you can call each command
# 	directly by name (instead of prefixing "tb." to each).
# 	However, this could create conflicts if another module
# 	with functions or global variables of the same name is
# 	loaded after Timebank, thus overriding Timebank's own.
#
# 	To start using Timebank, you first need a bank file.
# 	To create one using the default values, simply run
# 	"setup()" with no parameters in the parenthesis. You
# 	can change the settings later by using "change()".
#
# 	After creating the bank, add items using
# 	"debit(<time>, <tag>)".  To see your debits, use
# 	"showt()".  Remember to enclose non-number values
# 	using quotes, or the program will error.
#
# 	If the program does error during operation, don't fret --
# 	if your bank has been corrupted, you can restore it by
# 	replacing it with it's backup, <bankname> + '~'. Then
# 	report the bug at the Timebank github repository so it's
# 	fixed at the next release. :)
#
# 	If you have any complaints or suggestions, feel free to
# 	email me and although I may not answer all, I will surely
# 	read them.
#
# Written by Christen Gottschlich. Created July of 2012.
# For questions and/or bugs, email christen.got@gmail.com.
#
########################################


CHANGELOG
0.2	 --	Function "showt()" now can show percentages of tags,
		with better display format.

0.3  --	Fixed bug[0] by creating function "change()" which
		edits settings one at a time. Added "setup()" for
		easy bank initialization. "bank" file now uses five
		rows for settings, the last two being sleep time per
		sleep scope and estimated time slept for scope.
		Fixed bug[1], making sleep estimate truly accurate.

0.4	 --	Created functions "start()" and "end()", which when
		used together, record time for an activity and return
		the value. Fixed a runtime error where changing the
		unit leaves out writing of the estimated sleep time.



timebank.py
import os
# Conversion dictionary for scopes
s = {"d": 1, "w": 7, "m": 30}
# Conversion dictionary for display units
u = {"m": 1, "h": 60, "d": 60*24}
# Path of bank file
path = "bank"

# Configures a (new or existing) bank.
# scope := scope of entire bank, options are "d", "w", "m"
# unit := display unit, options are "m", "h", "d", "b"
# sleep := time spent sleeping within sleep scope (in hours)
# s_scope := scope of time sleeping, options are same as scope
# c_path := custom path.
# All amounts of time are stored as minutes and converted when
# necessary.
def config(scope="w", unit="h", sleep=8, s_scope="d"):

	# Calculate value for sleep
	s_total = sleep * s[scope] / s[s_scope] * 60
	# So that it's writable
	s_total = str(s_total)

	# Save to bank (or make new one)
	global path
	try: os.rename(path, path + '~')
	except: of = open(path + '~', "w")
	nf = open(path, 'w')
	of = open(path + "~", 'r')
	nf.write(unit + "\n"), nf.write(scope + "\n")
	nf.write(s_scope + "\n"), nf.write(s_total + "\n")
	lines = of.readlines()
	for i in range(4, len(lines)):
		nf.write(lines[i])
	nf.close()
	of.close()
	os.remove(path + "~")

# Returns a number equaling the time left available in the bank.
def avail():
	global path
	with open(path, "r") as f:
		lines = f.readlines()
		i_num = []
		i_name = []
		i_total = int(lines[3])
		for i in range(4, len(lines)):
			item = lines[i].split(":", 1)
			i_num.append(item[0])
			i_name.append(item[1])
			i_total = i_total + int(item[0])
		global s
		max_t = int(s[lines[1].rstrip('\n')]) * 24 * 60
		global u
		print (max_t - i_total) / u[lines[0].rstrip('\n')]

# Adds a debit to the bank.  If no tag is specified, then
# "nc" = no category, is supplied.  A time must be supplied.
def debit(time, tag="nc"):
	global path
	with open(path, "a") as f:
		f.write(str(time) + ':' + tag + '\n')

# Shows the amount of hours spent on input tag.  If no tag
# is specified, all tag totals are shown.  form determines
# the measurement of each tag; "unit" uses default unit,
# "perc" uses percentages.
def showtime(tag="all", form="unit"):
	global path
	with open(path, "r") as f:
		lines = f.readlines()
		tags = []
		times = []
		for i in range(4, len(lines)):
			lines[i] = lines[i].rstrip('\n')
			item = lines[i].split(":", 1)
			item[0] = int(item[0])
			if item[1] in tags:
				times[tags.index(item[1])] += item[0]
			else:
				tags.append(item[1])
				times.append(item[0])
		print tags
		print times
		# TODO: Better printing formating, multiple tags



tb.py
import os
import time
# Conversion dictionary for scopes
s = {"d": 1, "w": 7, "m": 30}
# Conversion dictionary for display units
u = {"m": 1, "h": 60, "d": 60*24}
# Conversion dictionary for scope names
sn = {"d": "day", "w": "week", "m": "month"}
# Path of bank file
bank_path = "mybank"

# Creates the initial "bank" file and fills in 
# default values (unless custom ones are
# submitted.
# "unit" := Display unit for time amounts.  Options
# are "m" (minute), "h" (hour), "d" (day), "b" (.beat).
# "scope" := Current span of time in frame. Options
# are "d" (day), "w" (week), "m" (month).
# "sleep" := Amount of sleep per "s_scope". Used
# in calculating the sleep estimate for the scope.
# Use 0 if you want to manually add/track sleep at
# each occurrence.
# "s_sleep" := Scope of sleep
def setup(scope="w", unit="h", sleep=8, s_scope="d"):
	# Save to bank (or make new one)
	try: open(bank_path, "r")
	except:
		print "Bank already exists! Use individual config"
		print "functions or delete old bank."
	else:
		with open(bank_path, "w") as f:
			t_sleep = sleep * s[scope] / s[s_scope] * 60
			f.write(unit + '\n'), f.write(scope + '\n'),
			f.write(s_scope + '\n'), f.write(str(sleep) + '\n'),
			f.write(str(t_sleep) + '\n')

# Changes individual settings. See "setup()" for details
# about settings.
# NOTE: If program crashes executing this command (for
# example, when forgetting quotes around a char value),
# make sure to restore the bank with the backup file
# (usually bank~) before running any more commands, to
# prevent bank content loss.
def change(setting, value):

	# See if bank exists
	try: open(bank_path, "r")
	except: print "No bank exists to change."

	# Perform change
	os.rename(bank_path, bank_path + '~')
	with open(bank_path + '~', "r") as of:
		with open(bank_path, "w") as nf:
			if setting == "unit":
				nf.write(str(value) + '\n')
				of.seek(2)
				is_scope = 3
			else: nf.write(of.readline())
			if setting == "scope":
				nf.write(str(value) + '\n')
				of.seek(4)
				is_scope = 2
			else: nf.write(of.readline())
			if setting == "s_scope":
				nf.write(str(value) + '\n')
				of.seek(6)
				is_scope = 1
			else: nf.write(of.readline())
			if setting == "sleep":
				nf.write(str(value) + '\n')
				of.seek(8)
				is_scope = 0
			else: nf.write(of.readline())
			nf.write(str(update(bank_path + '~', value, is_scope)) + '\n')

			# Copy the items, if they exist
			of.seek(0)
			lines = of.readlines()
			if len(lines) > 5:
				for i in range(5, len(lines)): 
					nf.write(lines[i])

	# Remove backup file (comment out to disable removal)
	os.remove(bank_path + '~')

# Updates sleep estimate.  Sleep estimate is used in
# predicting time remaining in bank spent awake. If
# you want all estimates of time remaining to not
# subtract estimated sleep, make this value 0.
# (You also have the option of manually
# managing sleep using "debit()"/"credit()" functions.)
# "update()" is run automatically when "sleep" values
# are changed, and therefore should never really be
# invoked manually.
def update(pathto, value, is_scope):
	with open(pathto, "r+") as f:
		if is_scope == 3:
			unit = value
			f.seek(2)
			scope = f.read(1)
			f.seek(4)
			s_scope = f.read(1)
			f.seek(6)
			sleep = int(f.read(1))
		elif is_scope == 2:
			scope = value
			f.seek(4)
			s_scope = f.read(1)
			f.seek(6)
			sleep = int(f.read(1))
			f.seek(0)
			unit = f.read(1)
		elif is_scope:
			s_scope = value
			f.seek(6)
			sleep = int(f.read(1))
			f.seek(2)
			scope = f.read(1)
			f.seek(0)
			unit = f.read(1)
		else:
			sleep = int(value)
			f.seek(4)
			s_scope = f.read(1)
			f.seek(2)
			scope = f.read(1)
			f.seek(0)
			unit = f.read(1)
		t_sleep = sleep * s[scope] / s[s_scope] * 60
		return t_sleep

# Returns a number equaling the time left available in the bank.
def avail():
	with open(bank_path, "r") as f:
		lines = f.readlines()
		i_num = []
		i_name = []
		i_total = int(lines[3])
		for i in range(4, len(lines)):
			item = lines[i].split(":", 1)
			i_num.append(item[1])
			i_name.append(item[0])
			i_total = i_total + int(item[1])
		max_t = int(s[lines[1].rstrip('\n')]) * 24 * 60
		print (max_t - i_total) / u[lines[0].rstrip('\n')]

# Adds a debit to the bank.  If no tag is specified, then
# "nc" = no category, is supplied.  A time must be supplied.
def debit(time, tag="nc"):
	global unit
	with open(bank_path, "a+") as f:
		unit = f.read(1)
		f.write(tag + ':' + str(time * u[unit]) + '\n')

# Shows the amount of hours spent on input tag.  If no tag
# is specified, all tag totals are shown.  "form" determines
# the measurement of each tag; no input uses default unit,
# "%" uses percentages.
def showt(tag="all", form=""):

	#TODO Maybe insert try-catch for "doesn't exist" error.
	with open(bank_path, "r") as f:
		lines = f.readlines()
		tags = []
		unit = lines[0].rstrip('\n')
		scope = lines[1].rstrip('\n')
		times = []
		spent_t = 0
	
	# If there are no items in bank yet, no show.  Otherwise
	if len(lines) < 5:
		print "No data to display yet. Add some with debit()"
	else:
		for i in range(5, len(lines)):
			lines[i] = lines[i].rstrip('\n')
			item = lines[i].split(":", 1)
			item[1] = int(item[1])
			spent_t = spent_t + item[1]
			if item[0] in tags:
				times[tags.index(item[0])] += item[1]
			else:
				tags.append(item[0])
				times.append(item[1])
		# Calculate total time in scope
		total_t = s[scope] * 24 * 60
		print "This {0}".format(sn[scope])
		if form == "%":
			for i in range(len(tags)):
				perc = 100 * times[i] / total_t
				print "{0:20} : {1:10d}%".format(tags[i], perc)
			print "{0:20} : {1:10}%".format("Total", 100 * spent_t / total_t)
		elif form == "":
			for i in range(len(tags)):
				print "{0:20} : {1:10d}".format(tags[i], times[i] / u[unit])
			print "{0:20} : {1:10} /{2}".format("Total", spent_t / u[unit], total_t / u[unit])

# Not yet implemented.
def credit(time, tag, message=""):
	pass

# Starts tracking time for an activity.
def start(tag):
	# Create file for recording
	with open(tag, "w") as f:
		f.write(str(int(time.time())))

# Ends timing an activity, returns time spent.
def end(tag):
	# Reopen record file
	with open(tag, "r") as f:
		diff = int(time.time()) - int(f.readline())
		# Diff is in seconds, return in minutes
		return diff / 60

